etna/models/nn/utils.py

```diff
@@ -184,8 +198,13 @@ def create_inference_dataset(self, ts: TSDataset) -> TimeSeriesDataSet:
             for feature_name in self.time_varying_known_categoricals:
                 df_flat[feature_name] = df_flat[feature_name].astype(str)
 
+        # `TimeSeriesDataSet.from_parameters` in predict mode ignores `min_prediction_length`,
+        # and we can change prediction size only by changing `max_prediction_length`
+        dataset_params = deepcopy(self.pf_dataset_params)
+        dataset_params["max_prediction_length"] = horizon
```

Comment:
It seems it could change the behaviour.
Have you checked both results - before changing and after?

Reply:
I'll explain the core of the problem. The problem was that `max_prediction_length` is set during training. You can set `min_predicition_length`, but it is ignored during forecasting and set equal to `max_prediction_length`. It is how pf works.

It leads to the situation when you can't make a forecast on dataset with smaller horizon that was used during training. It expects to forecast `max_prediction_length` points.

About the identity of the results I'll write report below in this discussion.

Reply:
I have a script
```python
from pytorch_forecasting.data import GroupNormalizer
from lightning_fabric.utilities.seed import seed_everything

from etna.datasets import TSDataset
from etna.datasets import generate_ar_df
from etna.models.nn import DeepARModel, PytorchForecastingDatasetBuilder
from etna.pipeline import Pipeline


def main():
    # load data
    df = generate_ar_df(periods=100, n_segments=3, start_time="2020-01-01", freq="D", random_seed=0)
    ts = TSDataset(df=TSDataset.to_dataset(df), freq="D")

    # fit pipeline
    builder = PytorchForecastingDatasetBuilder(
        max_encoder_length=5,
        max_prediction_length=5,
        time_varying_known_reals=["time_idx"],
        time_varying_unknown_reals=["target"],
        target_normalizer=GroupNormalizer(groups=["segment"]),
    )
    model = DeepARModel(dataset_builder=builder, trainer_params=dict(max_epochs=5), lr=0.01)
    pipeline = Pipeline(model=model, horizon=5)
    seed_everything(0)
    pipeline.fit(ts)

    # forecast
    ts_forecast_1 = pipeline.forecast()
    print(ts_forecast_1.to_pandas(flatten=True))


if __name__ == "__main__":
    main()
```

On this branch the result is:
```
    timestamp    segment    target
0  2020-04-10  segment_0  5.302200
1  2020-04-11  segment_0  5.245512
2  2020-04-12  segment_0  5.072897
3  2020-04-13  segment_0  4.956703
4  2020-04-14  segment_0  4.994010
5  2020-04-10  segment_1  8.561027
6  2020-04-11  segment_1  8.674180
7  2020-04-12  segment_1  8.962281
8  2020-04-13  segment_1  8.377350
9  2020-04-14  segment_1  8.384933
10 2020-04-10  segment_2 -6.026866
11 2020-04-11  segment_2 -6.008684
12 2020-04-12  segment_2 -5.824715
13 2020-04-13  segment_2 -6.108476
14 2020-04-14  segment_2 -5.785951
```

The same result is the same on the current `master` branch.

Code behavior isn't the same for all cases, because never version fixes inference tests. But I think this example proves that for normal scenario we haven't change logic of `DeepARModel`.

Reply:
In this example both max_prediction_lenght and horizon are the same.
I guess difference could arise in case of horzion smaller than max_prediction_lenght in case of transformers ( maybe if there is a difference it's a bug of source library - we shouldn't have bug in case of causal transformers )

Reply:
Another experiment:
```python
from pytorch_forecasting.data import GroupNormalizer
from lightning_fabric.utilities.seed import seed_everything

from etna.datasets import TSDataset
from etna.datasets import generate_ar_df
from etna.models.nn import DeepARModel, PytorchForecastingDatasetBuilder


def main():
    # load data
    df = generate_ar_df(periods=100, n_segments=3, start_time="2020-01-01", freq="D", random_seed=0)
    ts = TSDataset(df=TSDataset.to_dataset(df), freq="D")

    # fit pipeline
    builder = PytorchForecastingDatasetBuilder(
        max_encoder_length=5,
        max_prediction_length=5,
        time_varying_known_reals=["time_idx"],
        time_varying_unknown_reals=["target"],
        target_normalizer=GroupNormalizer(groups=["segment"]),
    )
    model = DeepARModel(dataset_builder=builder, trainer_params=dict(max_epochs=5), lr=0.01)
    seed_everything(0)
    model.fit(ts)

    # forecast
    future = ts.make_future(future_steps=3, tail_steps=model.context_size)
    result = model.forecast(future, prediction_size=3)
    print(result.to_pandas(flatten=True))


if __name__ == "__main__":
    main()
```

Current master branch: fails with error `AssertionError: filters should not remove entries all entries - check encoder/decoder lengths and lags`.

This branch: works fine with result:
```
   timestamp    segment    target
0 2020-04-10  segment_0  5.302200
1 2020-04-11  segment_0  5.245512
2 2020-04-12  segment_0  5.072897
3 2020-04-10  segment_1  8.561027
4 2020-04-11  segment_1  8.674180
5 2020-04-12  segment_1  8.962281
6 2020-04-10  segment_2 -6.026866
7 2020-04-11  segment_2 -6.008684
8 2020-04-12  segment_2 -5.824716
```